contracts/cw20-base/src/allowances.rs

```diff
@@ -0,0 +1,359 @@
+                if a.expires.is_expired(block) {
+                    Err(ContractError::Expired {})
+                } else {
+                    // deduct the allowance if enough
+                    a.allowance = a
+                        .allowance
+                        .checked_sub(amount)
+                        .map_err(StdError::overflow)?;
+                    Ok(a)
+                }
+            }
+            None => Err(ContractError::NoAllowance {}),
+        }
+    };
+    ALLOWANCES.update(storage, (owner, spender), update_fn)?;
+    ALLOWANCES_SPENDER.update(storage, (spender, owner), update_fn)
+}
+
+impl Cw20Allowances for Cw20Base {
+    type Error = ContractError;
+
+    /// Allows spender to access an additional amount tokens from the owner's (env.sender) account.
+    /// If expires is Some(), overwrites current allowance expiration with this one.
+    fn increase_allowance(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        spender: String,
+        amount: Uint128,
+        expires: Option<Expiration>,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let spender_addr = deps.api.addr_validate(&spender)?;
+        if spender_addr == info.sender {
+            return Err(ContractError::CannotSetOwnAccount {});
+        }
+
+        let update_fn = |allow: Option<AllowanceResponse>| -> Result<_, _> {
+            let mut val = allow.unwrap_or_default();
+            if let Some(exp) = expires {
+                if exp.is_expired(&env.block) {
+                    return Err(ContractError::InvalidExpiration {});
+                }
+                val.expires = exp;
+            }
+            val.allowance += amount;
+            Ok(val)
+        };
+        ALLOWANCES.update(deps.storage, (&info.sender, &spender_addr), update_fn)?;
+        ALLOWANCES_SPENDER.update(deps.storage, (&spender_addr, &info.sender), update_fn)?;
+
+        let res = Response::new().add_attributes(vec![
+            attr("action", "increase_allowance"),
+            attr("owner", info.sender),
+            attr("spender", spender),
+            attr("amount", amount),
+        ]);
+        Ok(res)
+    }
+
+    /// Lowers the spender's access of tokens from the owner's (env.sender) account by amount.
+    /// If expires is Some(), overwrites current allowance expiration with this one.
+    fn decrease_allowance(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        spender: String,
+        amount: Uint128,
+        expires: Option<Expiration>,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let spender_addr = deps.api.addr_validate(&spender)?;
+        if spender_addr == info.sender {
+            return Err(ContractError::CannotSetOwnAccount {});
+        }
+
+        let key = (&info.sender, &spender_addr);
+
+        fn reverse<'a>(t: (&'a Addr, &'a Addr)) -> (&'a Addr, &'a Addr) {
+            (t.1, t.0)
+        }
```

Comment:
So what is the point of this function instead of writing `|(a, b)| (b, a)` inline?